File: dashpilot-ios/dashpilot/Models/ManifestLoader.swift
import Foundation

struct ManifestLoader {

    // Loads manifest.json for each dashboard ID from the app bundle.
    // Expects files at web-{id}/manifest.json in Copy Bundle Resources.
    // Missing or unparseable manifests are logged and skipped; the rest load normally.
    static func loadFromBundle(dashboardIds: [String]) -> [String: DashboardManifest] {
        var result: [String: DashboardManifest] = [:]
        for id in dashboardIds {
            let subdirectory = "web-\(id)"
            guard let url = Bundle.main.url(
                forResource: "manifest",
                withExtension: "json",
                subdirectory: subdirectory
            ) else {
                print("[DashPilot] Missing manifest for '\(id)' — add \(subdirectory)/manifest.json to the Xcode target")
                continue
            }
            do {
                let json = try String(contentsOf: url, encoding: .utf8)
                let manifest = try DashboardManifest.parse(json)
                guard manifest.id == id else {
                    print("[DashPilot] Manifest at \(subdirectory)/manifest.json declares id '\(manifest.id)' but expected '\(id)'")
                    continue
                }
                result[id] = manifest
            } catch {
                print("[DashPilot] Failed to parse manifest for '\(id)': \(error)")
            }
        }
        return result
    }
}

File: dashpilot-android/app/src/main/java/com/softwiredtech/dashpilot/datamodel/dash/ManifestLoader.kt
package com.softwiredtech.dashpilot.datamodel.dash

import android.content.Context
import android.util.Log
import java.io.IOException

object ManifestLoader {
    fun loadFromAssets(context: Context, dashboardIds: List<String>): Map<String, DashboardManifest> {
        val result = mutableMapOf<String, DashboardManifest>()
        for (id in dashboardIds) {
            val path = "web-$id/manifest.json"
            try {
                val json = context.assets.open(path).bufferedReader().use { it.readText() }
                val manifest = DashboardManifest.parse(json)
                if (manifest.id != id) {
                    Log.w("DashPilot", "Manifest at assets/$path declares id '${manifest.id}' but expected '$id'")
                    continue
                }
                result[id] = manifest
            } catch (e: IOException) {
                Log.w("DashPilot", "Missing or unreadable manifest at assets/$path for dashboard '$id'", e)
            } catch (e: ManifestParseException) {
                Log.w("DashPilot", "Failed to parse manifest for dashboard '$id': ${e.message}", e)
            }
        }
        return result
    }
}
